### PR TITLE
reland [InlineAsm] wrap ConstraintCode in enum class NFC

### PR DESCRIPTION
**Pull request by nickdesaulniers**
reland [InlineAsm] wrap ConstraintCode in enum class NFC (#66003)

This reverts commit ee643b706be2b6bef9980b25cc9cc988dab94bb5.

Fix up build failures in targets I missed in #66003

Kept as 3 commits for reviewers to see better what's changed. Will squash when
merging.

- reland [InlineAsm] wrap ConstraintCode in enum class NFC (#66003)
- fix all the targets I missed in #66003
- fix off by one found by llvm/test/CodeGen/SystemZ/inline-asm-addr.ll

### CODE REVIEW

**Review by rnk — APPROVED** [review]
Looks good, I spot checked the Mips and SystemZ issues I saw locally. I think some stuff just got lost in your rebase attempt. I see the buildkite checks on Linux passed on this PR and the previous one.

**Comment by nickdesaulniers** [comment]
> Looks good, I spot checked the Mips and SystemZ issues I saw locally. I think some stuff just got lost in your rebase attempt. I see the buildkite checks on Linux passed on this PR and the previous one.

Thanks, looks like 142c89c381bccba1c688a90d7e9df91eaf78d4e3 is the culprit of that unittest failure on windows.

File: llvm/include/llvm/IR/InlineAsm.h [hunk 1]
```diff
@@ -375,11 +373,14 @@ class InlineAsm final : public Value {
...
 
-    // TODO: convert to enum?
-    unsigned getMemoryConstraintID() const {
+    ConstraintCode getMemoryConstraintID() const {
       assert((isMemKind() || isFuncKind()) &&
              "Not expected mem or function flag!");
-      return getData();
+      uint32_t D = getData();
+      assert(D <= static_cast<uint32_t>(ConstraintCode::Max) &&
+             D >= static_cast<uint32_t>(ConstraintCode::Unknown) &&
```
Comment (danilaml) [inline llvm/include/llvm/IR/InlineAsm.h:381]: Note that this triggers `type-limits` warning (if it's enabled) on GCC 7 and 8.
>error: comparison of unsigned expression >= 0 is always true

Don't know if there is a nice way to silence it and if we want to keep this check.

Comment (nickdesaulniers) [reply llvm/include/llvm/IR/InlineAsm.h:381]: Thanks for the report.

I'm about to delete this newly added code in https://github.com/llvm/llvm-project/pull/66297.

**Pull request merged**